### PR TITLE
New Resource: `azurerm_monitor_data_collection_rule`

### LINKED ISSUE #9679: Support for Azure Monitor Data Collection Rules
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

With the new Azure Monitor Guest Health agent extension, there is a condition that the enabling of this agent creates a required Data Collection Rule in Azure Monitor for VMInsights for this to work.

Please can we support the `Microsoft.Insights/dataCollectionRules` namespace as per the Azure API: https://docs.microsoft.com/en-us/rest/api/monitor/datacollectionrules

### New or Affected Resource(s)

* azurerm_monitor_data_collection_rule

### Potential Terraform Configuration

```hcl
resource "azurerm_monitor_data_collection_rule" "example" {
name                            = dataCollectionRuleName
resource_group_name             = resourceGroupName
location                        = location

settings = <<SETTINGS
  {
  "properties": {
    "dataSources": {
      "performanceCounters": [
        {
          "name": "cloudTeamCoreCounters",
          "streams": [
            "Microsoft-Perf"
          ],
          "scheduledTransferPeriod": "PT1M",
          "samplingFrequencyInSeconds": 15,
          "counterSpecifiers": [
            "\\Processor(_Total)\\% Processor Time",
            "\\Memory\\Committed Bytes",
            "\\LogicalDisk(_Total)\\Free Megabytes",
            "\\PhysicalDisk(_Total)\\Avg. Disk Queue Length"
          ]
        },
        {
          "name": "appTeamExtraCounters",
          "streams": [
            "Microsoft-Perf"
          ],
          "scheduledTransferPeriod": "PT5M",
          "samplingFrequencyInSeconds": 30,
          "counterSpecifiers": [
            "\\Process(_Total)\\Thread Count"
          ]
        }
      ],
      "windowsEventLogs": [
        {
          "name": "cloudSecurityTeamEvents",
          "streams": [
            "Microsoft-WindowsEvent"
          ],
          "scheduledTransferPeriod": "PT1M",
          "xPathQueries": [
            "Security!"
          ]
        },
        {
          "name": "appTeam1AppEvents",
          "streams": [
            "Microsoft-WindowsEvent"
          ],
          "scheduledTransferPeriod": "PT5M",
          "xPathQueries": [
            "System![System[(Level = 1 or Level = 2 or Level = 3)]]",
            "Application!*[System[(Level = 1 or Level = 2 or Level = 3)]]"
          ]
        }
      ],
      "syslog": [
        {
          "name": "cronSyslog",
          "streams": [
            "Microsoft-Syslog"
          ],
          "facilityNames": [
            "cron"
          ],
          "logLevels": [
            "Debug",
            "Critical",
            "Emergency"
          ]
        },
        {
          "name": "syslogBase",
          "streams": [
            "Microsoft-Syslog"
          ],
          "facilityNames": [
            "syslog"
          ],
          "logLevels": [
            "Alert",
            "Critical",
            "Emergency"
          ]
        }
      ]
    },
    "destinations": {
      "logAnalytics": [
        {
          "workspaceResourceId": "/subscriptions/<subscription_id>/resourceGroups/<resource_group>/providers/Microsoft.OperationalInsights/workspaces/centralTeamWorkspace",
          "name": "centralWorkspace"
        }
      ]
    },
    "dataFlows": [
      {
        "streams": [
          "Microsoft-Perf",
          "Microsoft-Syslog",
          "Microsoft-WindowsEvent"
        ],
        "destinations": [
          "centralWorkspace"
        ]
      }
    ]
  }
SETTINGS
}
```

### References

* https://docs.microsoft.com/en-us/rest/api/monitor/datacollectionrules

* #0000

### PR DESCRIPTION
**Pull request by teowa**
Support for new resource `azurerm_monitor_data_collection_rule`.
Reference: 
1. [Data Collection Rule](https://docs.microsoft.com/en-us/azure/azure-monitor/essentials/data-collection-rule-overview) 
2. [REST API ](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/monitor/resource-manager/Microsoft.Insights/stable/2021-04-01/dataCollectionRules_API.json).
---
```
TF_ACC=1 go test -v ./internal/services/monitor -run=TestAccMonitorDataCollectionRule_ -timeout 60m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccMonitorDataCollectionRule_basic
=== PAUSE TestAccMonitorDataCollectionRule_basic
=== RUN   TestAccMonitorDataCollectionRule_requiresImport
=== PAUSE TestAccMonitorDataCollectionRule_requiresImport
=== RUN   TestAccMonitorDataCollectionRule_update
=== PAUSE TestAccMonitorDataCollectionRule_update
=== RUN   TestAccMonitorDataCollectionRule_complete
=== PAUSE TestAccMonitorDataCollectionRule_complete
=== CONT  TestAccMonitorDataCollectionRule_basic
=== CONT  TestAccMonitorDataCollectionRule_update
=== CONT  TestAccMonitorDataCollectionRule_complete
=== CONT  TestAccMonitorDataCollectionRule_requiresImport
--- PASS: TestAccMonitorDataCollectionRule_basic (193.64s)
--- PASS: TestAccMonitorDataCollectionRule_requiresImport (205.91s)
--- PASS: TestAccMonitorDataCollectionRule_complete (211.10s)
--- PASS: TestAccMonitorDataCollectionRule_update (457.91s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/monitor       457.969s
```

(fixes #9679)

### CODE REVIEW

**Comment by tombuildsstuff** [comment]
@teowa can you please update this to use the SDK here: https://github.com/hashicorp/go-azure-sdk/tree/main/resource-manager/insights/2021-04-01 - rather than embedding this within this repository? Thanks!

File: internal/services/monitor/monitor_data_collection_rule_resource.go [hunk 1]
```diff
@@ -0,0 +1,892 @@
...
+			if err != nil && !response.WasNotFound(resp.HttpResponse) {
+				return fmt.Errorf("deleting %s: %+v", *id, err)
+			}
+			return nil
+		},
+		Timeout: 30 * time.Minute,
+	}
+}
+
+func expandStringTypeSlice[T ~string](input []string) *[]T {
```
Comment (mbfrahry) [inline internal/services/monitor/monitor_data_collection_rule_resource.go:559]: We're looking at how generics will fit into the provider so we're not quite ready to accept something like this yet. I appreciate the effort put into this but to merge this resource, we'll have to use the current way of mapping types.

**Review by mbfrahry — CHANGES_REQUESTED** [review]
Hi @teowa. Thanks for taking the time to write this PR out. It looks good but generics aren't something we're looking to add to the provider quite yet. Once that is removed, we should be good to get this in

File: internal/services/monitor/monitor_data_collection_rule_resource.go (hunk 1 above)
Comment (teowa) [reply internal/services/monitor/monitor_data_collection_rule_resource.go]: Thank @mbfrahry for reviewing. Despite the convenience, the use of generics also brings other risks, so it makes sense for not using generics in the provider now. I have changed the code and rerun the tests. Please continue review.
```
--- PASS: TestAccMonitorDataCollectionRule_complete (157.58s)
--- PASS: TestAccMonitorDataCollectionRule_basic (186.86s)
--- PASS: TestAccMonitorDataCollectionRule_requiresImport (222.00s)
--- PASS: TestAccMonitorDataCollectionRule_update (458.63s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/monitor       459.741s

```

File: internal/services/monitor/monitor_data_collection_rule_resource.go [hunk 2]
```diff
@@ -0,0 +1,1017 @@
...
+									Required: true,
+									MinItems: 1,
+									Elem: &pluginsdk.Schema{
+										Type: pluginsdk.TypeString,
+										ValidateFunc: validation.StringInSlice(
+											datacollectionrules.PossibleValuesForKnownExtensionDataSourceStreams(),
+											false),
+									},
+								},
+								"extension_setting": {
```
Comment (katbyte) [inline internal/services/monitor/monitor_data_collection_rule_resource.go:208]: should this be
```suggestion
								"extension_json": {
```

**Review by katbyte — CHANGES_REQUESTED** [review]

**Comment by teowa** [comment]
Find a previous PR of Data Collection Rule: https://github.com/hashicorp/terraform-provider-azurerm/pull/16704  made by @ms-henglu , thanks to him, he has confirmed with service team that receiving of`Microsoft-WindowsEvent` streams in Log Analytic Workspace must be explicitly enabled (using `azurerm_log_analytics_solution` ) in the future API version(2021-09-01-preview or later). To be consistent,  I have added `azurerm_log_analytics_solution` in acctest and docs like what does in [heng's PR](https://github.com/hashicorp/terraform-provider-azurerm/pull/16704/files#diff-e05d0e8161312a6764a35ff8fdffadd96f9489a9203f08d7e496fe6eedc2cf9bR129)

**Review by mbfrahry — APPROVED** [review]
LGTM!

**Pull request merged**